### ISSUE TITLE
undefined | string property of object incorrectly narrowed to string in union type

### ISSUE THREAD

**Issue by everhardt** [microsoft/TypeScript#53264]
# Bug Report

### 🔎 Search Terms

union undefined incorrectly narrowed

### 🕗 Version & Regression Information

This is the behavior in every version I tried (back to 3.3.3), and I reviewed the FAQ for entries about bugs, undefined and unions.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.9.5#code/DYUwLgBAhgXBDOYBOBLAdgcwgHwgVzQBMQAzdEQgbgChqwBPABxAgCEB9ARggF4IBvaHESpMAGggBjOAHIwIRDImFZ9EFAAWMiAF8aDZm3YAmXgKmz5inXSYtWZjt1wdjNagHoPEMBpTwEDQB7PGBCKSCAW0ZgKHQfDSgwT28AWghGJCDmJAZoQJCwiDQgyBLIEAAPf2SvCHSgpAysnLyoAtDwgCMWEXQsKCJm7JBc+ghw-whI-3h+6kkgtEQILrgHPkEoCWkIOQUwGT1aFBIIAAoZKG14roBKAWoIZ4g6nskoPHgWILPfKdmeBAEi6ECmaCgSCyAHcKD4gkZuBpRsCINC-KB8lMCMQyGg4YNwpMAuVmgoQGhki8IssgqAAHTAIIYc73Gg2IA)

### 💻 Code

```ts
let a: string | undefined;

type B_1 = { a: string, c: 'test', d: 'yeah' };
type B_2 = { c: 'test'}
type B = B_1 | B_2;

// this should complain that
// - property a should not not exist
// - or property a should be string and property d is missing
const b: B = { a, c: 'test'};

if ('a' in b) {
    // because of this issue, b is narrowed to B_1 here, while a is undefined and d is not present
    console.log(b);
}

```

### 🙁 Actual behavior

`const b: B = { a, c: 'test'};` does not throw a TS error.

### 🙂 Expected behavior

`const b: B = { a, c: 'test'};` throws a TS error.

**Comment by MartinJohns** [comment]
The assignment is allowed, because `{ a, c: 'test' }` is a valid `B` (specifically, it's a valid `B_2`). Excess property checks is not performed for union types, see #51873.

This unfortunately leads to an intentional unsoundness using the `in` operator, see #34975.

**Comment by everhardt** [comment]
Thanks for your quick reply! I think this issue can then be closed as it doesn't add much to the two issues you mentioned. Sorry for wasting your time.
